File: java/src/main/java/sk/epostak/sdk/resources/DocumentsResource.java
package sk.epostak.sdk.resources;

import sk.epostak.sdk.HttpClient;
import sk.epostak.sdk.models.*;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Send and receive documents via Peppol.
 * <p>
 * Provides operations for sending invoices, checking delivery status,
 * downloading PDF/UBL representations, and responding to received invoices.
 * <p>
 * Access via {@code client.documents()}.
 *
 * <pre>{@code
 * // Send an invoice
 * SendDocumentResponse result = client.documents().send(
 *     SendDocumentRequest.builder("0245:1234567890")
 *         .invoiceNumber("FV-2026-001")
 *         .issueDate("2026-04-04")
 *         .items(List.of(new SendDocumentRequest.LineItem(
 *             "Consulting", 10, 50, 23)))
 *         .build());
 *
 * // Check status
 * DocumentStatusResponse status = client.documents().status(result.documentId());
 * }</pre>
 */
public final class DocumentsResource {

    private final HttpClient http;
    private final InboxResource inbox;

    /**
     * Creates a new documents resource.
     *
     * @param http the HTTP client used for API communication
     */
    public DocumentsResource(HttpClient http) {
        this.http = http;
        this.inbox = new InboxResource(http);
    }

    /**
     * Access received (inbound) documents.
     *
     * @return the inbox sub-resource
     */
    public InboxResource inbox() {
        return inbox;
    }

    /**
     * Get a document by ID.
     *
     * <pre>{@code
     * Document doc = client.documents().get("doc_abc123");
     * System.out.println(doc.getStatus()); // "DELIVERED"
     * }</pre>
     *
     * @param id the document UUID
     * @return the full document object
     * @throws sk.epostak.sdk.EPostakException if the document is not found (404) or the request fails
     */
    public Document get(String id) {
        return http.get("/documents/" + HttpClient.encode(id), Document.class);
    }

    /**
     * Update a draft document. Only documents in {@code DRAFT} status can be updated.
     *
     * @param id      the document UUID
     * @param request the fields to update (all optional)
     * @return the updated document
     * @throws sk.epostak.sdk.EPostakException if the document is not a draft or the request fails
     */
    public Document update(String id, UpdateDocumentRequest request) {
        return http.patch("/documents/" + HttpClient.encode(id), request, Document.class);
    }

    /**
     * Send a document via Peppol. Supports both JSON mode (structured data) and
     * XML mode (pre-built UBL).
     *
     * <pre>{@code
     * SendDocumentResponse result = client.documents().send(
     *     SendDocumentRequest.builder("0245:1234567890")
     *         .invoiceNumber("FV-2026-001")
     *         .issueDate("2026-04-04")
     *         .dueDate("2026-04-18")
     *         .items(List.of(new SendDocumentRequest.LineItem(
     *             "Consulting services", 10, 150.0, 23)))
     *         .build());
     * }</pre>
     *
     * @param request the send document request
     * @return the response containing document ID, message ID, and initial status
     * @throws sk.epostak.sdk.EPostakException if validation fails or the request fails
     */
    public SendDocumentResponse send(SendDocumentRequest request) {
        return http.post("/documents/send", request, SendDocumentResponse.class);
    }

    /**
     * Get full document status including status history timeline.
     *
     * @param id the document UUID
     * @return the status response with history entries
     * @throws sk.epostak.sdk.EPostakException if the document is not found or the request fails
     */
    public DocumentStatusResponse status(String id) {
        return http.get("/documents/" + HttpClient.encode(id) + "/status", DocumentStatusResponse.class);
    }

    /**
     * Get delivery evidence for a sent document, including AS4 receipt and
     * Message Level Response (MLR) data.
     *
     * @param id the document UUID
     * @return the evidence response with receipts and response documents
     * @throws sk.epostak.sdk.EPostakException if the document is not found or the request fails
     */
    public DocumentEvidenceResponse evidence(String id) {
        return http.get("/documents/" + HttpClient.encode(id) + "/evidence", DocumentEvidenceResponse.class);
    }

    /**
     * Download the document rendered as PDF bytes.
     *
     * <pre>{@code
     * byte[] pdfBytes = client.documents().pdf("doc_abc123");
     * Files.write(Path.of("invoice.pdf"), pdfBytes);
     * }</pre>
     *
     * @param id the document UUID
     * @return raw PDF bytes
     * @throws sk.epostak.sdk.EPostakException if the document is not found or the request fails
     */
    public byte[] pdf(String id) {
        return http.getBytes("/documents/" + HttpClient.encode(id) + "/pdf");
    }

    /**
     * Download the UBL XML representation of the document.
     *
     * @param id the document UUID
     * @return the UBL XML as a string
     * @throws sk.epostak.sdk.EPostakException if the document is not found or the request fails
     */
    public String ubl(String id) {
        return http.getString("/documents/" + HttpClient.encode(id) + "/ubl");
    }

    /**
     * Download the signed AS4 envelope for this document from the 10-year WORM
     * archive (S3 Object Lock COMPLIANCE). Returns the raw multipart AS4 payload
     * exactly as it was transmitted on the Peppol network — signed, timestamped,
     * and tamper-evident. Usable as dispute evidence or for regulatory retention.
     * <p>
     * Availability: {@code api-enterprise} plan only; other plans get
     * {@link sk.epostak.sdk.EPostakException} with HTTP 403.
     * <p>
     * Very recently sent documents may briefly return 404 until the archival
     * cron picks them up — retry after a short delay rather than treating the
     * first 404 as permanent.
     *
     * <pre>{@code
     * byte[] as4 = client.documents().envelope("doc_abc123");
     * Files.write(Path.of("doc_abc123.as4"), as4);
     * }</pre>
     *
     * @param id the document UUID
     * @return raw AS4 envelope bytes (multipart payload)
     * @throws sk.epostak.sdk.EPostakException if the document is not found, the envelope
     *         is not yet archived, the plan is not {@code api-enterprise}, or the request fails
     */
    public byte[] envelope(String id) {
        return http.getBytes("/documents/" + HttpClient.encode(id) + "/envelope");
    }

    /**
     * Send an invoice response for a received document. Status must be one of
     * the seven UBL response codes: {@code AB}, {@code IP}, {@code UQ},
     * {@code CA}, {@code RE}, {@code AP}, {@code PD}. See
     * {@link InvoiceRespondRequest} for the full legend.
     *
     * <pre>{@code
     * InvoiceRespondResponse resp = client.documents().respond("doc_abc123",
     *     new InvoiceRespondRequest("AP", "Accepted, thank you"));
     * }</pre>
     *
     * @param id      the document UUID of the received invoice
     * @param request the response containing status code and optional note (max 500 chars)
     * @return the respond confirmation with timestamp and dispatch status
     * @throws sk.epostak.sdk.EPostakException if the document is not found or the request fails
     */
    public InvoiceRespondResponse respond(String id, InvoiceRespondRequest request) {
        return http.post("/documents/" + HttpClient.encode(id) + "/respond", request, InvoiceRespondResponse.class);
    }

    /**
     * Validate a UBL XML document against Peppol BIS 3.0 rules without sending it.
     *
     * <pre>{@code
     * ValidationResult result = client.documents().validate(ublXml);
     * if (!result.valid()) {
     *     result.errors().forEach(e -> System.err.println(e.rule() + ": " + e.message()));
     * }
     * }</pre>
     *
     * @param ublXml the UBL 2.1 XML document to validate
     * @return the validation result with per-rule errors and warnings
     * @throws sk.epostak.sdk.EPostakException if the request fails or the validator is unavailable
     */
    public ValidationResult validate(String ublXml) {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("format", "ubl");
        body.put("document", ublXml);
        return http.post("/documents/validate", body, ValidationResult.class);
    }

    /**
     * Validate a JSON-mode invoice body by generating UBL server-side and running
     * full Peppol BIS 3.0 validation against it.
     *
     * @param jsonDocument the invoice document (same shape as {@link SendDocumentRequest}, as a map/POJO)
     * @return the validation result with per-rule errors and warnings
     * @throws sk.epostak.sdk.EPostakException if the request fails or the validator is unavailable
     */
    public ValidationResult validateJson(Object jsonDocument) {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("format", "json");
        body.put("document", jsonDocument);
        return http.post("/documents/validate", body, ValidationResult.class);
    }

    /**
     * Check whether a receiver supports Peppol document delivery (preflight check).
     *
     * <pre>{@code
     * PreflightResult result = client.documents().preflight("0245:1234567890");
     * if (result.canSend()) {
     *     // Safe to send
     * }
     * }</pre>
     *
     * @param receiverPeppolId receiver Peppol ID, e.g. {@code "0245:1234567890"}
     * @param documentType     optional UBL document type identifier to check specific capability
     * @return the preflight result indicating registration and capability support
     * @throws sk.epostak.sdk.EPostakException if the request fails
     */
    public PreflightResult preflight(String receiverPeppolId, String documentType) {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("receiverPeppolId", receiverPeppolId);
        if (documentType != null) {
            body.put("documentType", documentType);
        }
        return http.post("/documents/preflight", body, PreflightResult.class);
    }

    /**
     * Check whether a receiver supports Peppol document delivery, for any document type.
     *
     * @param receiverPeppolId receiver Peppol ID, e.g. {@code "0245:1234567890"}
     * @return the preflight result indicating registration status
     * @throws sk.epostak.sdk.EPostakException if the request fails
     */
    public PreflightResult preflight(String receiverPeppolId) {
        return preflight(receiverPeppolId, null);
    }

    /**
     * Convert between JSON and UBL XML document representations.
     *
     * <pre>{@code
     * // JSON to UBL
     * ConvertResult ubl = client.documents().convert("json", "ubl", jsonData);
     *
     * // UBL to JSON
     * ConvertResult json = client.documents().convert("ubl", "json", xmlString);
     * }</pre>
     *
     * @param inputFormat  the format of {@code document}: {@code "json"} or {@code "ubl"}
     * @param outputFormat the desired output format: {@code "ubl"} or {@code "json"}
     * @param document     the document to convert — a {@code Map}/POJO for JSON input,
     *                     or a UBL XML {@code String} for UBL input
     * @return the conversion result containing {@code outputFormat}, {@code document}, and {@code warnings}
     * @throws sk.epostak.sdk.EPostakException if the conversion fails or the request fails
     */
    public ConvertResult convert(String inputFormat, String outputFormat, Object document) {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("input_format", inputFormat);
        body.put("output_format", outputFormat);
        body.put("document", document);
        return http.post("/documents/convert", body, ConvertResult.class);
    }

    /**
     * Send multiple documents in a single call. Items are processed independently:
     * a failure on one item does not abort the batch. The response contains a
     * per-item status in the same order as the request.
     *
     * <pre>{@code
     * BatchSendResponse resp = client.documents().sendBatch(List.of(
     *     new BatchSendRequest.BatchItem(
     *         SendDocumentRequest.builder("0245:12345678")
     *             .invoiceNumber("FV-2026-001").build()),
     *     new BatchSendRequest.BatchItem(
     *         SendDocumentRequest.builder("0245:87654321")
     *             .invoiceNumber("FV-2026-002").build(), "idem-002")
     * ));
     * System.out.println(resp.succeeded() + "/" + resp.total() + " sent");
     * }</pre>
     *
     * @param items the batch items to send; max 50 per call, 20 MB total body size
     * @return the batch send response with per-item results
     * @throws sk.epostak.sdk.EPostakException if the request fails
     */
    public BatchSendResponse sendBatch(List<BatchSendRequest.BatchItem> items) {
        return http.post("/documents/send/batch", new BatchSendRequest(items), BatchSendResponse.class);
    }

    /**
     * Parse a UBL XML invoice into structured JSON without persisting or sending it.
     * The returned shape matches the JSON side of {@link #convert(String, String, Object)}
     * with {@code output_format=json}.
     *
     * <pre>{@code
     * ParsedInvoice parsed = client.documents().parse(ublXmlString);
     * System.out.println(parsed.document().get("invoice_number"));
     * }</pre>
     *
     * @param xml the UBL 2.1 XML document
     * @return the parsed invoice and any non-fatal warnings
     * @throws sk.epostak.sdk.EPostakException if the XML cannot be parsed or the request fails
     */
    public ParsedInvoice parse(String xml) {
        return http.postRaw("/documents/parse", xml, "application/xml", ParsedInvoice.class);
    }

    /**
     * Mark the processing state of an inbound document.
     * <p>
     * {@code state} must be one of {@code "delivered"}, {@code "processed"},
     * {@code "failed"}, or {@code "read"}. The optional {@code note} is recorded
     * alongside the transition (e.g. failure reason).
     *
     * <pre>{@code
     * MarkResponse r = client.documents().mark("doc_abc123", "processed", "Booked to ledger");
     * System.out.println(r.status()); // overall doc status after marking
     * }</pre>
     *
     * @param id    the document UUID to mark
     * @param state target state: {@code "delivered"}, {@code "processed"}, {@code "failed"}, or {@code "read"}
     * @param note  optional note (max 500 chars), or {@code null}
     * @return the mark response with updated timestamps
     * @throws sk.epostak.sdk.EPostakException if the document is not found or the state transition is invalid
     */
    public MarkResponse mark(String id, String state, String note) {
        return http.post(
                "/documents/" + HttpClient.encode(id) + "/mark",
                new MarkRequest(state, note),
                MarkResponse.class
        );
    }

    /**
     * Mark the processing state of an inbound document, without a note.
     *
     * @param id    the document UUID to mark
     * @param state target state: {@code "delivered"}, {@code "processed"}, {@code "failed"}, or {@code "read"}
     * @return the mark response with updated timestamps
     * @throws sk.epostak.sdk.EPostakException if the document is not found or the state transition is invalid
     */
    public MarkResponse mark(String id, String state) {
        return mark(id, state, null);
    }
}

File: dotnet/src/EPostak/Resources/DocumentsResource.cs
using EPostak.Models;

namespace EPostak.Resources;

/// <summary>
/// Send, receive, and manage e-invoicing documents via the Peppol network.
/// Provides operations for the full document lifecycle: sending, receiving,
/// status tracking, validation, and format conversion.
/// </summary>
public sealed class DocumentsResource
{
    private readonly HttpRequestor _http;

    /// <summary>Access received (inbound) documents from the Peppol network.</summary>
    public InboxResource Inbox { get; }

    internal DocumentsResource(HttpRequestor http)
    {
        _http = http;
        Inbox = new InboxResource(http);
    }

    /// <summary>
    /// Retrieve a document by its unique identifier.
    /// Returns the full document including supplier/customer details, line items, and totals.
    /// </summary>
    /// <param name="id">The document UUID.</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>The complete document with all metadata, parties, lines, and totals.</returns>
    /// <example>
    /// <code>
    /// var doc = await client.Documents.GetAsync("doc_abc123");
    /// Console.WriteLine($"{doc.Number}: {doc.Totals.WithVat} {doc.Currency}");
    /// </code>
    /// </example>
    public Task<Document> GetAsync(string id, CancellationToken ct = default)
        => _http.RequestAsync<Document>(HttpMethod.Get, $"/documents/{Uri.EscapeDataString(id)}", ct);

    /// <summary>
    /// Update a draft document before sending. Only documents in draft status can be updated.
    /// Any fields left null in the request will remain unchanged.
    /// </summary>
    /// <param name="id">The document UUID of the draft to update.</param>
    /// <param name="request">Fields to update on the draft document.</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>The updated document with all current values.</returns>
    /// <example>
    /// <code>
    /// var updated = await client.Documents.UpdateAsync("doc_abc123", new UpdateDocumentRequest
    /// {
    ///     DueDate = "2026-05-01",
    ///     Note = "Updated payment terms"
    /// });
    /// </code>
    /// </example>
    public Task<Document> UpdateAsync(string id, UpdateDocumentRequest request, CancellationToken ct = default)
        => _http.RequestAsync<Document>(HttpMethod.Patch, $"/documents/{Uri.EscapeDataString(id)}", request, ct);

    /// <summary>
    /// Send an e-invoice via the Peppol network to a registered recipient.
    /// You can either provide structured JSON data (with line items) or raw UBL XML.
    /// The document is validated, converted to UBL if needed, and transmitted via AS4.
    /// </summary>
    /// <remarks>
    /// <para>
    /// <strong>Document types (<c>DocType</c>).</strong> Accepted values:
    /// <c>"invoice"</c>, <c>"credit_note"</c>, <c>"correction"</c>,
    /// <c>"self_billing"</c>, <c>"reverse_charge"</c>,
    /// <c>"self_billing_credit_note"</c>. Defaults to <c>"invoice"</c> when omitted.
    /// </para>
    /// <para>
    /// <strong>Supplier-party pinning (XML mode).</strong> When submitting raw
    /// UBL via <see cref="SendDocumentRequest.Xml"/>, the server pins the seller
    /// identity (Name, IČO, IČ DPH, Postal Address, Legal Entity name) to the
    /// authenticated firm. Caller-supplied values in
    /// <c>cac:AccountingSupplierParty/cac:Party</c> are silently overwritten
    /// before forwarding to Peppol. The Peppol <c>EndpointID</c> is the only
    /// supplier-party field still validated against the firm's registered
    /// Peppol ID; mismatched EndpointIDs are rejected with HTTP 422. BG-24
    /// attachments, line items, payment terms, and custom note fields are
    /// preserved as-is. For self-billing document types (UBL typecodes
    /// <c>261</c>/<c>389</c>), the customer party (which is the authenticated
    /// firm) is rewritten instead.
    /// </para>
    /// </remarks>
    /// <param name="request">The invoice data including receiver Peppol ID and line items or UBL XML.</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>The document ID, Peppol message ID, and initial delivery status.</returns>
    /// <example>
    /// <code>
    /// var result = await client.Documents.SendAsync(new SendDocumentRequest
    /// {
    ///     ReceiverPeppolId = "0192:12345678",
    ///     InvoiceNumber = "FV-2026-001",
    ///     IssueDate = "2026-04-11",
    ///     DueDate = "2026-05-11",
    ///     Currency = "EUR",
    ///     Items = new List&lt;LineItem&gt;
    ///     {
    ///         new() { Description = "Consulting", Quantity = 10, UnitPrice = 100m, VatRate = 23m, Unit = "HUR" }
    ///     }
    /// });
    /// Console.WriteLine($"Sent: {result.DocumentId}, Status: {result.Status}");
    /// </code>
    /// </example>
    public Task<SendDocumentResponse> SendAsync(SendDocumentRequest request, CancellationToken ct = default)
        => _http.RequestAsync<SendDocumentResponse>(HttpMethod.Post, "/documents/send", request, ct);

    /// <summary>
    /// Get the current delivery status and full status history of a document.
    /// Use this to track whether a sent invoice has been delivered, acknowledged, or responded to.
    /// </summary>
    /// <param name="id">The document UUID.</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>Current status, status history timeline, and optional validation/response details.</returns>
    /// <example>
    /// <code>
    /// var status = await client.Documents.StatusAsync("doc_abc123");
    /// Console.WriteLine($"Current: {status.Status}");
    /// foreach (var entry in status.StatusHistory)
    ///     Console.WriteLine($"  {entry.Timestamp}: {entry.Status}");
    /// </code>
    /// </example>
    public Task<DocumentStatusResponse> StatusAsync(string id, CancellationToken ct = default)
        => _http.RequestAsync<DocumentStatusResponse>(HttpMethod.Get, $"/documents/{Uri.EscapeDataString(id)}/status", ct);

    /// <summary>
    /// Get delivery evidence for a sent document. This includes the AS4 receipt from
    /// the receiver's access point, the Message Level Response (MLR), and any
    /// invoice response (accept/reject/query) from the recipient.
    /// </summary>
    /// <param name="id">The document UUID.</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>Available evidence: AS4 receipt, MLR document, and invoice response with timestamps.</returns>
    /// <example>
    /// <code>
    /// var evidence = await client.Documents.EvidenceAsync("doc_abc123");
    /// if (evidence.InvoiceResponse is not null)
    ///     Console.WriteLine($"Response: {evidence.InvoiceResponse.Status}");
    /// </code>
    /// </example>
    public Task<DocumentEvidenceResponse> EvidenceAsync(string id, CancellationToken ct = default)
        => _http.RequestAsync<DocumentEvidenceResponse>(HttpMethod.Get, $"/documents/{Uri.EscapeDataString(id)}/evidence", ct);

    /// <summary>
    /// Download the PDF visualization of a document. The PDF is generated from the
    /// UBL XML and includes a human-readable invoice layout.
    /// </summary>
    /// <param name="id">The document UUID.</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>The PDF file contents as a byte array.</returns>
    /// <example>
    /// <code>
    /// byte[] pdf = await client.Documents.PdfAsync("doc_abc123");
    /// await File.WriteAllBytesAsync("invoice.pdf", pdf);
    /// </code>
    /// </example>
    public Task<byte[]> PdfAsync(string id, CancellationToken ct = default)
        => _http.RequestBytesAsync(HttpMethod.Get, $"/documents/{Uri.EscapeDataString(id)}/pdf", ct);

    /// <summary>
    /// Download the signed AS4 envelope of a document from the 10-year WORM archive
    /// (S3 Object Lock COMPLIANCE mode). Returns the raw multipart AS4 payload exactly
    /// as it was transmitted on the Peppol network — signed, timestamped, and
    /// tamper-evident. Available on api-enterprise plan during an active contract;
    /// every document that ever flowed through the AP is retrievable for 10 years.
    /// </summary>
    /// <remarks>
    /// Returns <c>404</c> when the invoice doesn't exist, doesn't belong to the firm,
    /// or the envelope hasn't been archived yet (the archive cron runs on a short
    /// interval, so brand-new documents may briefly miss — retry shortly).
    /// </remarks>
    /// <param name="id">The document UUID.</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>The signed AS4 envelope bytes.</returns>
    /// <example>
    /// <code>
    /// byte[] envelope = await client.Documents.EnvelopeAsync("doc_abc123");
    /// await File.WriteAllBytesAsync($"{id}.as4", envelope);
    /// </code>
    /// </example>
    public Task<byte[]> EnvelopeAsync(string id, CancellationToken ct = default)
        => _http.RequestBytesAsync(HttpMethod.Get, $"/documents/{Uri.EscapeDataString(id)}/envelope", ct);

    /// <summary>
    /// Download the UBL XML source of a document. This is the canonical Peppol BIS 3.0
    /// representation that was sent or received over the network.
    /// </summary>
    /// <param name="id">The document UUID.</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>The UBL 2.1 XML document as a string.</returns>
    /// <example>
    /// <code>
    /// string ubl = await client.Documents.UblAsync("doc_abc123");
    /// File.WriteAllText("invoice.xml", ubl);
    /// </code>
    /// </example>
    public Task<string> UblAsync(string id, CancellationToken ct = default)
        => _http.RequestStringAsync(HttpMethod.Get, $"/documents/{Uri.EscapeDataString(id)}/ubl", ct);

    /// <summary>
    /// Respond to a received document with an Invoice Response. Sends a Peppol BIS
    /// Invoice Response 3.0 message back to the original sender. The
    /// <see cref="InvoiceRespondRequest.Status"/> must be one of the seven UNCL4343
    /// codes: <c>AB</c> (accepted billing), <c>IP</c> (in process), <c>UQ</c> (under
    /// query), <c>CA</c> (conditionally accepted), <c>RE</c> (rejected), <c>AP</c>
    /// (accepted), or <c>PD</c> (paid).
    /// </summary>
    /// <param name="id">The document UUID of the received invoice to respond to.</param>
    /// <param name="request">The response status and optional note (max 500 chars).</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>Confirmation of the response including document ID and timestamp.</returns>
    /// <example>
    /// <code>
    /// var response = await client.Documents.RespondAsync("doc_abc123", new InvoiceRespondRequest
    /// {
    ///     Status = InvoiceResponseCode.AP,
    ///     Note = "Invoice accepted for payment"
    /// });
    /// </code>
    /// </example>
    public Task<InvoiceRespondResponse> RespondAsync(string id, InvoiceRespondRequest request, CancellationToken ct = default)
        => _http.RequestAsync<InvoiceRespondResponse>(HttpMethod.Post, $"/documents/{Uri.EscapeDataString(id)}/respond", request, ct);

    /// <summary>
    /// Validate a document without sending it. Checks the invoice data against
    /// Peppol BIS 3.0 rules and Slovak e-invoicing requirements. For JSON mode,
    /// also returns the generated UBL XML.
    /// </summary>
    /// <param name="request">The invoice data to validate (same format as <see cref="SendAsync"/>).</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>Validation result with valid/invalid flag, warnings, and generated UBL XML (for JSON mode).</returns>
    /// <example>
    /// <code>
    /// var result = await client.Documents.ValidateAsync(new SendDocumentRequest
    /// {
    ///     ReceiverPeppolId = "0192:12345678",
    ///     Items = new List&lt;LineItem&gt;
    ///     {
    ///         new() { Description = "Item 1", Quantity = 1, UnitPrice = 50m, VatRate = 23m }
    ///     }
    /// });
    /// if (!result.Valid)
    ///     Console.WriteLine($"Warnings: {string.Join(", ", result.Warnings)}");
    /// </code>
    /// </example>
    public Task<ValidationResult> ValidateAsync(SendDocumentRequest request, CancellationToken ct = default)
        => _http.RequestAsync<ValidationResult>(HttpMethod.Post, "/documents/validate", request, ct);

    /// <summary>
    /// Check whether a receiver is registered on the Peppol network and can accept a given
    /// document type. Performs an SMP (Service Metadata Publisher) lookup without sending anything.
    /// Use this before sending to verify the recipient exists.
    /// </summary>
    /// <param name="request">The receiver Peppol ID and optional document type to check.</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>Registration status, document type support, and SMP URL for the receiver.</returns>
    /// <example>
    /// <code>
    /// var check = await client.Documents.PreflightAsync(new PreflightRequest
    /// {
    ///     ReceiverPeppolId = "0192:12345678"
    /// });
    /// if (check.Registered &amp;&amp; check.SupportsDocumentType)
    ///     Console.WriteLine("Receiver is ready to accept invoices");
    /// </code>
    /// </example>
    public Task<PreflightResult> PreflightAsync(PreflightRequest request, CancellationToken ct = default)
        => _http.RequestAsync<PreflightResult>(HttpMethod.Post, "/documents/preflight", request, ct);

    /// <summary>
    /// Convert between JSON and UBL XML document formats. Useful for previewing the UBL
    /// that would be generated from JSON data, or parsing received UBL into structured JSON.
    /// </summary>
    /// <param name="request">The input/output formats and the source document (JSON object or UBL XML string).</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>The converted output in the target format, plus any non-fatal warnings.</returns>
    /// <example>
    /// <code>
    /// var result = await client.Documents.ConvertAsync(new ConvertRequest
    /// {
    ///     InputFormat = ConvertInputFormat.Ubl,
    ///     OutputFormat = ConvertOutputFormat.Json,
    ///     Document = ublXmlString
    /// });
    /// </code>
    /// </example>
    public Task<ConvertResult> ConvertAsync(ConvertRequest request, CancellationToken ct = default)
        => _http.RequestAsync<ConvertResult>(HttpMethod.Post, "/documents/convert", request, ct);

    /// <summary>
    /// Send multiple documents in a single call. Items are processed independently --
    /// a failure on one item does not abort the batch. The response contains a per-item
    /// status in the same order as the request.
    /// </summary>
    /// <param name="items">Items to send. Max 100 per call.</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>Batch totals (total/succeeded/failed) and per-item results.</returns>
    /// <example>
    /// <code>
    /// var response = await client.Documents.SendBatchAsync(new List&lt;BatchSendItem&gt;
    /// {
    ///     new() { Document = new SendDocumentRequest { ReceiverPeppolId = "0192:12345678", ... } },
    ///     new() { Document = new SendDocumentRequest { ReceiverPeppolId = "0192:87654321", ... },
    ///             IdempotencyKey = "batch-002" }
    /// });
    /// Console.WriteLine($"{response.Succeeded}/{response.Total} sent");
    /// </code>
    /// </example>
    public Task<BatchSendResponse> SendBatchAsync(List<BatchSendItem> items, CancellationToken ct = default)
        => _http.RequestAsync<BatchSendResponse>(
            HttpMethod.Post,
            "/documents/send/batch",
            new BatchSendRequest { Items = items },
            ct);

    /// <summary>
    /// Parse a UBL XML invoice into structured JSON without persisting or sending it.
    /// The returned shape matches the JSON side of <see cref="ConvertAsync"/>
    /// with <see cref="ConvertOutputFormat.Json"/>.
    /// </summary>
    /// <param name="xml">UBL 2.1 XML document.</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>The parsed invoice and any non-fatal warnings.</returns>
    /// <example>
    /// <code>
    /// var parsed = await client.Documents.ParseAsync(ublXmlString);
    /// var json = (JsonElement)parsed.Document;
    /// Console.WriteLine(json.GetProperty("invoiceNumber").GetString());
    /// </code>
    /// </example>
    public Task<ParsedInvoice> ParseAsync(string xml, CancellationToken ct = default)
        => _http.RequestRawAsync<ParsedInvoice>(HttpMethod.Post, "/documents/parse", xml, "application/xml", ct);

    /// <summary>
    /// Mark the processing state of an inbound document. <paramref name="state"/> must be
    /// one of <c>delivered</c>, <c>processed</c>, <c>failed</c>, or <c>read</c>. The optional
    /// <paramref name="note"/> is recorded alongside the transition (e.g. failure reason).
    /// </summary>
    /// <param name="id">Document UUID to mark.</param>
    /// <param name="state">Target state: <c>delivered</c>, <c>processed</c>, <c>failed</c>, or <c>read</c>.</param>
    /// <param name="note">Optional note (max 500 chars).</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>The mark response with updated timestamps and overall document status.</returns>
    /// <example>
    /// <code>
    /// var marked = await client.Documents.MarkAsync("doc_abc123", "processed", "Booked to ledger");
    /// Console.WriteLine($"Status now: {marked.Status}");
    /// </code>
    /// </example>
    public Task<MarkResponse> MarkAsync(string id, string state, string? note = null, CancellationToken ct = default)
        => _http.RequestAsync<MarkResponse>(
            HttpMethod.Post,
            $"/documents/{Uri.EscapeDataString(id)}/mark",
            new MarkRequest { State = state, Note = note },
            ct);

    /// <summary>
    /// List outbound (sent) documents with optional filtering and pagination.
    /// </summary>
    /// <param name="params">Optional query parameters for filtering and pagination.</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>Paginated list of outbound documents.</returns>
    /// <example>
    /// <code>
    /// var outbox = await client.Documents.OutboxAsync(new OutboxParams { Limit = 50 });
    /// foreach (var doc in outbox.Documents)
    ///     Console.WriteLine($"{doc.Number}: {doc.Status}");
    /// </code>
    /// </example>
    public Task<OutboxListResponse> OutboxAsync(OutboxParams? @params = null, CancellationToken ct = default)
    {
        var qs = HttpRequestor.BuildQuery(
            ("offset", @params?.Offset?.ToString()),
            ("limit", @params?.Limit?.ToString()),
            ("status", @params?.Status),
            ("peppolMessageId", @params?.PeppolMessageId),
            ("since", @params?.Since));
        return _http.RequestAsync<OutboxListResponse>(HttpMethod.Get, $"/documents/outbox{qs}", ct);
    }

    /// <summary>
    /// Get all Invoice Responses received for a document (accept/reject/query from the buyer).
    /// </summary>
    /// <param name="id">Document UUID.</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>List of invoice response records with status codes and timestamps.</returns>
    /// <example>
    /// <code>
    /// var result = await client.Documents.ResponsesAsync("doc_abc123");
    /// foreach (var r in result.Responses)
    ///     Console.WriteLine($"{r.ResponseCode}: {r.CreatedAt}");
    /// </code>
    /// </example>
    public Task<InvoiceResponsesListResponse> ResponsesAsync(string id, CancellationToken ct = default)
        => _http.RequestAsync<InvoiceResponsesListResponse>(HttpMethod.Get, $"/documents/{Uri.EscapeDataString(id)}/responses", ct);

    /// <summary>
    /// Get the audit trail events for a document. Uses cursor-based pagination.
    /// </summary>
    /// <param name="id">Document UUID.</param>
    /// <param name="params">Optional pagination parameters (limit, cursor).</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>Ordered list of audit events with next cursor.</returns>
    /// <example>
    /// <code>
    /// var result = await client.Documents.EventsAsync("doc_abc123");
    /// foreach (var evt in result.Events)
    ///     Console.WriteLine($"{evt.OccurredAt}: {evt.EventType} ({evt.Actor})");
    /// </code>
    /// </example>
    public Task<DocumentEventsResponse> EventsAsync(string id, DocumentEventsParams? @params = null, CancellationToken ct = default)
    {
        var qs = HttpRequestor.BuildQuery(
            ("limit", @params?.Limit?.ToString()),
            ("cursor", @params?.Cursor));
        return _http.RequestAsync<DocumentEventsResponse>(HttpMethod.Get, $"/documents/{Uri.EscapeDataString(id)}/events{qs}", ct);
    }
}
